File: packages/guard/src/core/evaluate.ts
import type {
  CompiledPolicy,
  Decision,
  Finding,
  NormalizedToolEvent,
  RegexMatcher,
  RuleMatch
} from "../types/index.js";
import { getByPath } from "../utils/path.js";
import { severityToDecision } from "./decision.js";

function toRegex(input: RegexMatcher): RegExp {
  return new RegExp(input.regex, input.flags ?? "i");
}

function asArray(value?: string | string[]): string[] {
  if (!value) return [];
  if (Array.isArray(value)) return value;
  return [value];
}

function resolveAction(severityAction: Decision | undefined, fallbackSeverity: Decision): Decision {
  return severityAction ?? fallbackSeverity;
}

function matchesText(text: string, matcher?: RegexMatcher): boolean {
  if (!matcher) return true;
  return toRegex(matcher).test(text);
}

function matchesIntent(intent: string, matcher?: RegexMatcher): boolean {
  if (!matcher) return true;
  return toRegex(matcher).test(intent);
}

function matchesTool(toolName: string, matcher?: string | string[]): boolean {
  if (!matcher) return true;
  const allowed = asArray(matcher).map((item) => item.toLowerCase());
  return allowed.includes(toolName.toLowerCase());
}

function matchesArg(args: unknown, matcher: RuleMatch["arg"]): boolean {
  if (!matcher) return true;

  const value = getByPath(args, matcher.path);

  if (matcher.eq !== undefined) {
    return value === matcher.eq;
  }

  if (matcher.regex) {
    const re = new RegExp(matcher.regex, matcher.flags ?? "i");
    return re.test(String(value ?? ""));
  }

  return value !== undefined;
}

function matchesDestination(domain: string | undefined, matcher: RuleMatch["destination"]): boolean {
  if (!matcher?.domain) return true;

  const expected = asArray(matcher.domain).map((item) => item.toLowerCase());
  if (!domain) return false;

  return expected.some((candidate) => domain === candidate || domain.endsWith(`.${candidate}`));
}

function matchedKeys(event: NormalizedToolEvent, match: RuleMatch): string[] {
  const keys: string[] = [];

  if (match.tool && matchesTool(event.toolName, match.tool)) keys.push("tool");
  if (match.text && matchesText(event.text, match.text)) keys.push("text");
  if (match.intent && matchesIntent(event.intent, match.intent)) keys.push("intent");
  if (match.arg && matchesArg(event.args, match.arg)) keys.push(`arg:${match.arg.path}`);
  if (match.destination && matchesDestination(event.destinationDomain, match.destination)) keys.push("destination");

  return keys;
}

export function evaluatePolicy(event: NormalizedToolEvent, policy: CompiledPolicy): Finding[] {
  const findings: Finding[] = [];

  for (const rule of policy.rules) {
    const match = rule.match;

    const criteriaPass =
      matchesTool(event.toolName, match.tool) &&
      matchesText(event.text, match.text) &&
      matchesIntent(event.intent, match.intent) &&
      matchesArg(event.args, match.arg) &&
      matchesDestination(event.destinationDomain, match.destination);

    if (!criteriaPass) continue;

    const fallbackDecision = severityToDecision(rule.severity);
    const findingAction = resolveAction(rule.action, fallbackDecision);

    findings.push({
      eventId: event.eventId,
      ruleId: rule.id,
      title: rule.title ?? rule.id,
      category: rule.category,
      severity: rule.severity,
      action: findingAction,
      confidence: rule.confidence ?? policy.defaults.confidence ?? 0.75,
      why: rule.why,
      suggestion: rule.suggestion,
      matchedOn: matchedKeys(event, match)
    });
  }

  return findings;
}

File: packages/python/src/twire_guard/evaluate.py
from __future__ import annotations

from typing import Any

from .decision import severity_to_decision
from .utils import compile_regex, get_by_path


def _as_array(value: Any) -> list[str]:
    if value is None:
        return []
    if isinstance(value, list):
        return [str(item) for item in value]
    return [str(value)]


def _resolve_action(rule_action: str | None, fallback_action: str) -> str:
    return rule_action or fallback_action


def _matches_text(text: str, matcher: dict[str, Any] | None) -> bool:
    if matcher is None:
        return True
    return bool(compile_regex(str(matcher["regex"]), matcher.get("flags"), default_insensitive=True).search(text))


def _matches_intent(intent: str, matcher: dict[str, Any] | None) -> bool:
    if matcher is None:
        return True
    return bool(compile_regex(str(matcher["regex"]), matcher.get("flags"), default_insensitive=True).search(intent))


def _matches_tool(tool_name: str, matcher: str | list[str] | None) -> bool:
    if matcher is None:
        return True
    allowed = [item.lower() for item in _as_array(matcher)]
    return tool_name.lower() in allowed


def _matches_arg(args: Any, matcher: dict[str, Any] | None) -> bool:
    if matcher is None:
        return True

    value = get_by_path(args, str(matcher.get("path", "")))

    if "eq" in matcher:
        return value == matcher.get("eq")

    regex = matcher.get("regex")
    if isinstance(regex, str):
        return bool(
            compile_regex(regex, matcher.get("flags"), default_insensitive=True).search(
                "" if value is None else str(value)
            )
        )

    return value is not None


def _matches_destination(domain: str | None, matcher: dict[str, Any] | None) -> bool:
    if matcher is None or matcher.get("domain") is None:
        return True

    expected = [item.lower() for item in _as_array(matcher.get("domain"))]
    if not domain:
        return False

    lower = domain.lower()
    return any(lower == candidate or lower.endswith(f".{candidate}") for candidate in expected)


def _matched_keys(event: dict[str, Any], match: dict[str, Any]) -> list[str]:
    keys: list[str] = []

    if match.get("tool") is not None and _matches_tool(str(event["tool_name"]), match.get("tool")):
        keys.append("tool")
    if match.get("text") is not None and _matches_text(str(event["text"]), match.get("text")):
        keys.append("text")
    if match.get("intent") is not None and _matches_intent(str(event["intent"]), match.get("intent")):
        keys.append("intent")
    if match.get("arg") is not None and _matches_arg(event.get("args"), match.get("arg")):
        keys.append(f"arg:{match['arg']['path']}")
    if match.get("destination") is not None and _matches_destination(event.get("destination_domain"), match.get("destination")):
        keys.append("destination")

    return keys


def evaluate_policy(event: dict[str, Any], policy: dict[str, Any]) -> list[dict[str, Any]]:
    findings: list[dict[str, Any]] = []

    for rule in policy.get("rules", []):
        match = rule.get("match", {})

        criteria_pass = (
            _matches_tool(str(event["tool_name"]), match.get("tool"))
            and _matches_text(str(event.get("text", "")), match.get("text"))
            and _matches_intent(str(event.get("intent", "")), match.get("intent"))
            and _matches_arg(event.get("args"), match.get("arg"))
            and _matches_destination(event.get("destination_domain"), match.get("destination"))
        )

        if not criteria_pass:
            continue

        fallback_decision = severity_to_decision(str(rule.get("severity", "low")))
        finding_action = _resolve_action(rule.get("action"), fallback_decision)

        defaults = policy.get("defaults", {}) if isinstance(policy.get("defaults"), dict) else {}
        confidence = rule.get("confidence")
        if not isinstance(confidence, (int, float)):
            default_confidence = defaults.get("confidence")
            confidence = float(default_confidence) if isinstance(default_confidence, (int, float)) else 0.75

        findings.append(
            {
                "event_id": event["event_id"],
                "rule_id": rule["id"],
                "title": rule.get("title") or rule["id"],
                "category": rule["category"],
                "severity": rule["severity"],
                "action": finding_action,
                "confidence": confidence,
                "why": rule["why"],
                "suggestion": rule["suggestion"],
                "matched_on": _matched_keys(event, match),
            }
        )

    return findings
